packages/astro/src/env/vite-plugin-env-virtual-mod.ts

```diff
@@ -1,28 +1,117 @@
-import type { Plugin } from 'vite';
+import { loadEnv, type Plugin } from 'vite';
 import type { AstroSettings } from '../@types/astro.js';
 import type { Logger } from '../core/logger/core.js';
+import {
+	ENV_TYPES_FILE,
+	RESOLVED_VIRTUAL_CLIENT_MODULE_ID,
+	VIRTUAL_CLIENT_MODULE_ID,
+} from './constants.js';
+import type { EnvSchema } from './schema.js';
+import { validateEnvVariable } from './validators.js';
+import { AstroError, AstroErrorData } from '../core/errors/index.js';
+import type fsMod from 'node:fs';
+import { fileURLToPath } from 'node:url';
 
 interface AstroEnvVirtualModPluginParams {
 	settings: AstroSettings;
 	logger: Logger;
+	mode: 'dev' | 'build' | string;
+	fs: typeof fsMod;
 }
 
 export function astroEnvVirtualModPlugin({
 	settings,
 	logger,
+	mode,
+	fs,
 }: AstroEnvVirtualModPluginParams): Plugin | undefined {
 	if (!settings.config.experimental.env) {
 		return;
 	}
 
 	logger.warn('env', 'This feature is experimental. TODO:');
 
-	const { schema } = settings.config.experimental.env;
-	// TODO: client / public
+	const schema = settings.config.experimental.env.schema ?? {};
+	const loadedEnv = loadEnv(
+		mode === 'dev' ? 'development' : 'production',
+		fileURLToPath(settings.config.root),
+		''
+	);
+
+	const { clientContent, clientDts } = handleClientModule({ schema, loadedEnv });
+	handleDts({ settings, fs, content: clientDts });
+
 	// TODO: server / public
 	// TODO: server / secret
 	return {
 		name: 'astro-env-virtual-mod-plugin',
 		enforce: 'pre',
+		resolveId(id) {
+			if (id === VIRTUAL_CLIENT_MODULE_ID) {
+				return RESOLVED_VIRTUAL_CLIENT_MODULE_ID;
+			}
+		},
+		load(id) {
+			if (id === RESOLVED_VIRTUAL_CLIENT_MODULE_ID) {
+				return clientContent;
+			}
+		},
+	};
+}
+
+function handleDts({
+	content,
+	settings,
+	fs,
+}: {
+	content: string;
+	settings: AstroSettings;
+	fs: typeof fsMod;
+}) {
+	fs.mkdirSync(settings.dotAstroDir, { recursive: true });
+	fs.writeFileSync(new URL(ENV_TYPES_FILE, settings.dotAstroDir), content, 'utf-8');
```

Comment:
nit: It's a long shot, but I would wrap these in a `try/catch` in case we can't write files on disk